### PR TITLE
Added mkcloud step for assigning aliases to the nodes

### PR DESCRIPTION
**Pull request by jsuchome**
First part is based on user's input.
Aliases are needed for referencing the nodes in scenario files.

In next steps, we'll like to have some logic to assign aliases automatically.

### CODE REVIEW

**Comment by aspiers** [comment]
Still WIP, but marked by a label now.

**Comment by vuntz** [comment]
Won't that somehow conflict with the code setting the `dashboard` alias for the horizon node?

**Comment by jsuchome** [comment]
Do we need that dashboard alias? Maybe it could be removed.

dasboard alias is set at the end of proposal step, so currently it means, that some node will get its alias change again. This should not affect the ``batch`` step, which is not supposed to be used together with proposal

**Comment by aspiers** [comment]
In the HA case, the dashboard alias is somewhat broken, since it just points to the first node in the cluster running Horizon.  Also the `dashboard.[cgh]X.cloud.suse.de` A record is not guaranteed to point to the same node which has the `dashboard` alias.  So I think we can probably drop it.  Here's some context:

```
2015-08-10 11:56:04     aspiers dirk_: any reason why I shouldn't drop the qa_crowbarsetup.sh code which aliases one node to 'dashboard'? it doesn't really make any sense
2015-08-10 11:56:19     aspiers that alias makes it look like it's a dedicated dashboard node
2015-08-10 11:56:42     aspiers when in reality it could be running everything, or might not be the only dashboard node
2015-08-10 11:57:16     dirk_   aspiers: its the one node available via dashboard.[cgh]X.cloud.suse.de
2015-08-10 11:57:27     dirk_   its not the "dashboard" node. 
2015-08-10 11:57:36     dirk_   I agree that with HA it would ideally be the cluster vip
2015-08-10 11:58:41     aspiers are we talking about the same thing?
2015-08-10 11:59:37     aspiers how can external DNS point to the right IP when it's dynamically allocated?
2015-08-10 12:01:20     dirk_   magic
2015-08-10 12:01:23     dirk_   :-)
2015-08-10 12:01:42     Magic   At your service!
2015-08-10 12:01:46     aspiers lol
2015-08-10 12:01:56     aspiers Magic: please explain how you work :)
2015-08-10 12:02:07     Magic   By magic
2015-08-10 12:02:18     aspiers bah, you're as helpful as dirk_ 
2015-08-10 12:02:26     Magic   lol
2015-08-10 12:02:27     aspiers :)
2015-08-10 12:03:05     aspiers dirk_: explain the magic or it doesn't exist ;-p
2015-08-10 12:03:38     dirk_   aspiers: there is nat'ing in between, that is being resolved to the ndoe that has the dashboard name
2015-08-10 12:05:29     dirk_   aspiers: I agree that it doesn't make a lot of sense to preserve that right now with HA in mind, but it also doesn't hurt to keep it for now, and that way we could focus on the other open things..
2015-08-10 12:05:53     dirk_   its really from the times when there was only one controller node..
2015-08-10 12:07:27     aspiers ok
2015-08-10 12:08:27     aspiers I still have no idea how NAT can work on a dynamically allocated IP though, without querying crowbar
2015-08-10 12:08:45     aspiers or crowbar's DNS
2015-08-10 12:09:09     aspiers last time I checked, NAT was an L3 function
2015-08-10 12:11:49     aspiers dirk_: also btw hX.cloud.suse.de is completely missing from DNS anyway
2015-08-10 12:12:15     dirk_   aspiers: the Ip address is predictable
2015-08-10 12:12:35     dirk_   aspiers: yes, thats a known bug. now that jdsn is back, perhaps he could fix it ;-)
2015-08-10 12:14:04     jdsn    dirk_: yes I can
2015-08-10 12:14:48     jdsn    dirk_: I still need to see what has been changed during the last weeks .... then I hopefully find the new ranges
2015-08-10 12:14:52     aspiers dirk_: only predictable in the non-HA case when one of the allocated [cgh]X.cloud.suse.de networks is used. but I'll leave it for now anyway
2015-08-10 12:16:15     aspiers dirk_: no, this still doesn't make any sense. if the NAT hack is just relying on a predictable IP, then the dashboard alias in crowbar is totally redundant
2015-08-10 12:18:08     aspiers dirk_: redundant from that PoV at least. I can sort of see how it could be convenient for manual testing
2015-08-10 12:19:11     aspiers so AFAICS your statement "there is nat'ing in between, that is being resolved to the node that has the dashboard name" cannot be right
2015-08-10 12:21:47     dirk_   aspiers: ok. lets reword it to "the node which gets the dashboard alias is the one that gets the dashboard.cX nating assigned"
2015-08-10 12:21:59     dirk_   aspiers: as such it is convenient for a non-HA setup
2015-08-10 12:22:18     dirk_   aspiers: and 99.99999999999999% of all successful mkcloud runs are non-HA ;-)
2015-08-10 12:31:42     aspiers dirk_: actually 100% :-/
2015-08-10 12:33:12     aspiers dirk_: but that is still not an argument to keep the alias in crowbar, since the alias is irrelevant for that forwarding
2015-08-10 12:38:37     aspiers dirk_: anyway, I'll leave it there for now but it's good to understand why it's there and it sounds like nothing actually needs it ...
```

File: scripts/qa_crowbarsetup.sh [hunk 1]
```diff
@@ -2401,7 +2409,7 @@ function prepare_proposals()
 function set_dashboard_alias()
```
Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: so should this be removed completely?

File: scripts/mkcloud [hunk 2]
```diff
@@ -967,6 +974,7 @@ Steps:
     setupnodes:     create the nodes and let crowbar discover them
     instnodes:      allocate and install compute nodes
     proposal:       create and apply proposals for default setup
+    setup_aliases:  set aliases for the nodes (usually needed for batch step)
```
Comment (jdsn) [inline scripts/mkcloud]: Could we please call it crowbar_batch (at least not only batch).

Comment (jsuchome) [inline scripts/mkcloud]: the name of the step is ``batch``

File: scripts/qa_crowbarsetup.sh [hunk 3]
```diff
@@ -3421,6 +3436,95 @@ function onadmin_run_cct()
...
+            # 2. non-HA
+            # 1st node is controller by default (intended role is set by onadmin_allocate)
+            local controller=`get_all_discovered_nodes  | head -n1`
+            set_node_alias $controller "controller"
+            nodesavailable=`printf "%s\n" $nodesavailable | grep -iv $controller`
+
+            i=1
+            # storage nodes (cephN or swiftN) will exist based on deployceph/deployswift value
+            if [ -n "$deployceph" ] || [ -n "$deployswift" ] ; then
+                for node in `get_all_discovered_nodes | grep -v $controller | head -n2` ; do
```
Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: We always use 2 nodes for ceph or swift?

Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: For ceph, it's minimum, but swift can use just one node. In case you have just 2 nodes in total, this loop will take only one.

Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: But in general, the automatic logic is not perfect and I think never could be. User should provide aliases via want_node_aliases value, esp. if he/she wants to use them in scenario description.

**Comment by jsuchome** [comment]
This was tested with explicit setting of want_node_aliases, and also without it (both HA and non-HA case)

**Comment by jdsn** [comment]
There are many additions in the first commits that are removed or changed in the later commits. I recommend to squash changes at the same location if possible (but still maintain the one logical change per commit rule).

File: scripts/mkcloud (hunk 2 above)
Comment (jdsn) [inline scripts/mkcloud]: Ok, but I agree not in this PR.

File: scripts/qa_crowbarsetup.sh [hunk 4]
```diff
@@ -3421,10 +3441,97 @@ function onadmin_run_cct()
...
+            i=1
+            for node in $clusternodesnetwork ; do
+                set_node_alias $node "network$i"
+                i=$((i+1))
+            done
+            i=1
+            for node in $clusternodesservices ; do
+                set_node_alias $node "services$i"
+                i=$((i+1))
+            done
```
Comment (bmwiedemann) [inline scripts/qa_crowbarsetup.sh]: 3x code duplication - not so nice... would probably need an eval to handle the different variable names
```bash
for x in data network services ; do
  eval "for node in \$custernodes$x ; do
    set_node_alias \$node \"$x\$i\"
    ...
  done" ; 
done
```

File: scripts/qa_crowbarsetup.sh (hunk 1 above)
Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: I actually think it doesn't belong to this PR. And if the new step is used together with batch, proposal should not be called anyway, thus no ``set_dashboard_alias``

**Comment by jdsn** [comment]
Here are two squash candidates: https://github.com/jsuchome/automation/commit/3c0c62204f3f7607590b7eed4ec5613bc3aa668a

**Comment by jdsn** [comment]
Still 2 hunks in here
https://github.com/jsuchome/automation/commit/821c4da14cd8d5469ae070f7a3fd5f598578e21d
that should be squashed to the first commit.

**Comment by jdsn** [comment]
This is how it looks like right after the split:
https://github.com/jdsn/SUSE-Cloud--automation/commits/jdsn/pr/566

And this is the expected outcome (after merging the new commit to the first one):
https://github.com/jdsn/SUSE-Cloud--automation/commits/jdsn/pr/566-squashed

File: scripts/mkcloud (hunk 2 above)
Comment (aspiers) [inline scripts/mkcloud]: I prefer `batch` - less typing, and still everyone knows what it does.  Plus everything in `mkcloud` is related to Crowbar anyway, so the `crowbar_` prefix seems redundant to me.  Why do you want the longer version?

**Comment by jsuchome** [comment]
@jdsn thanks for help

File: scripts/qa_crowbarsetup.sh [hunk 5]
```diff
@@ -3492,10 +3512,91 @@ function onadmin_run_cct()
...
+                eval "cluster=\$clusternodes$clustername"
+                i=1
+                for node in $cluster ; do
+                    set_node_alias $node "$clustername$i"
+                    i=$((i+1))
+                done
+            done
+            i=1
+            for node in $unclustered_nodes ; do
+                set_node_alias_and_role $node "compute$i" compute
```
Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: I wouldn't set a role here; the function is really about setting aliases and I feel it's out of scope to set roles.

File: scripts/qa_crowbarsetup.sh [hunk 6]
```diff
@@ -3492,10 +3512,91 @@ function onadmin_run_cct()
...
+            # 1st node is controller by default (intended role is set by onadmin_allocate)
+            local controller=`get_all_discovered_nodes  | head -n1`
+            set_node_alias $controller "controller"
+            nodesavailable=`printf "%s\n" $nodesavailable | grep -iv $controller`
+
+            i=1
+            # storage nodes (cephN or swiftN) will exist based on deployceph/deployswift value
+            if [ -n "$deployceph" ] || [ -n "$deployswift" ] ; then
+                for node in `get_all_discovered_nodes | grep -v $controller | head -n2` ; do
+                    set_node_alias_and_role $node "storage$i" storage
```
Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: (same comment)

File: scripts/qa_crowbarsetup.sh [hunk 7]
```diff
@@ -3492,10 +3512,91 @@ function onadmin_run_cct()
...
+                    nodesavailable=`printf "%s\n" $nodesavailable | grep -iv $node`
+                    i=$((i+1))
+                done
+            fi
+
+            # Use computeN for the rest.
+            i=1
+            for node in $nodesavailable; do
+                set_node_alias_and_role $node "compute$i" compute
```
Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: (and again)

Alternatively, the function should be renamed to something else to make it clear it's not about just setting aliases.

File: scripts/qa_crowbarsetup.sh (hunk 5 above)
Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: I know it wasn't necessary, but I thought it's nice additional touch, when we are already trying to guess the roles anyway.
It might be handy if user wants to continue with some parts of deployment manually.

**Comment by bmwiedemann** [comment]
:+1:

**Comment by nkrinner** [comment]
+1

**Comment by vuntz** [comment]
+1
